rclpy/src/rclpy/wait_set.cpp

```diff
@@ -0,0 +1,295 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/error_handling.h>
+#include <rcl/rcl.h>
+#include <rcl/types.h>
+
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+extern "C"
+{
+#include "rclpy_common/common.h"
+}
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+#include "wait_set.hpp"
+
+namespace rclpy
+{
+/// Destructor for a wait set
+void
+_rclpy_destroy_wait_set(PyObject * pycapsule)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(PyCapsule_GetPointer(pycapsule, "rcl_wait_set_t"));
+
+  rcl_ret_t ret = rcl_wait_set_fini(wait_set);
+  if (ret != RCL_RET_OK) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level,
+      "Failed to fini wait set: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+
+  PyMem_Free(wait_set);
+}
+
+py::capsule
+get_zero_initialized_wait_set()
+{
+  auto deleter = [](rcl_wait_set_t * ptr) {PyMem_FREE(ptr);};
+  auto wait_set = std::unique_ptr<rcl_wait_set_t, decltype(deleter)>(
+    static_cast<rcl_wait_set_t *>(PyMem_Malloc(sizeof(rcl_wait_set_t))),
+    deleter);
```

Comment:
In the other pybind11 conversions, we've been throwing `bad_malloc` if `PyMem_Malloc` failed. Should we do the same here?

Reply:
throws `std::bad_alloc` if `PyMem_Malloc` fails in bb10dab5683597429488e5ba85c193edee66f998